tools/c7n_azure/c7n_azure/resources/vm.py

```diff
@@ -47,3 +48,43 @@ def __call__(self, i):
             i['instanceView'] = instance.serialize()
 
         return super(InstanceViewFilter, self).__call__(i['instanceView'])
+
+
+@VirtualMachine.action_registry.register('stop')
+class VmStopAction(BaseAction):
+
+    schema = type_schema('stop')
+
+    def process(self, vms):
+        for vm in vms:
+            self.manager.get_client().virtual_machines.power_off(vm['resourceGroup'], vm['name'])
```

Comment:
you're instantiating a new client on each VM which can definitely be in the order of magnitude of 1000's of vms aka thousands of sessions, better to move the client instantiation out of the loop. Same thing applies for start/restart/delete